app/views/feedback.py

```diff
@@ -59,22 +59,23 @@ def send_feedback():
 
         metadata = get_metadata(current_user)
 
-        message = convert_feedback(
-            escape(request.form.get('message')),
-            escape(request.form.get('name')),
-            escape(request.form.get('email')),
-            request.referrer or '',
-            metadata,
-            g.schema.json['survey_id'],
-        )
-
-        encrypted_message = encrypt(message, current_app.eq['key_store'], key_purpose=KEY_PURPOSE_SUBMISSION)
-        sent = current_app.eq['submitter'].send_message(encrypted_message,
-                                                        current_app.config['EQ_RABBITMQ_QUEUE_NAME'],
-                                                        metadata.get('tx_id'))
-
-        if not sent:
-            raise SubmissionFailedException()
+        if metadata:
```

Comment:
I wonder if the safest thing to do here would be to throw a `NoTokenException` when there is no metadata in the same way as I did in my last commit?

There's potential with this change for the user to be redirected to the thank you page even when the feedback is not submitted, which might be a bit confusing.

Reply:
Personally I don't think this is necessary. So with this change, if you post a feedback without filling any fields and without a valid session, this would evaluate line 58:
https://github.com/ONSdigital/eq-survey-runner/blob/8906c8219c69ad85cc347f7908891b7a4563fd56/app/views/feedback.py#L58

Then line 80:
https://github.com/ONSdigital/eq-survey-runner/blob/8906c8219c69ad85cc347f7908891b7a4563fd56/app/views/feedback.py#L80-L81

This would then redirect you to thank you page, from which the decorator `with_metadata_context` is called on the GET request, then on line 34, 35 of template helper it would evaluate false and raise a no token exception.
https://github.com/ONSdigital/eq-survey-runner/blob/8906c8219c69ad85cc347f7908891b7a4563fd56/app/helpers/template_helper.py#L34-L35

---

However, we could move line 60:
https://github.com/ONSdigital/eq-survey-runner/blob/8906c8219c69ad85cc347f7908891b7a4563fd56/app/views/feedback.py#L60 above the`if form.validate()` call and fail early to avoid the extra evaluations. So above form.validate we could do:

```python
metadata = get_metadata(current_user)
if not metadata:
    raise NoTokenException(401)
```

Hope that is clear, let me know if I've made no sense.

Reply:
It depends on what we want to display to the user. The current method redirects to the thank you page which then displays the session expired message. Raising the 401 in feedback.py displays the generic "Error 500 - something has gone wrong" message.

Personally I think the session expired message is clearer.

Reply:
Raising a 401 will return a session expired page as it has a custom handler.
https://github.com/ONSdigital/eq-survey-runner/blob/8906c8219c69ad85cc347f7908891b7a4563fd56/app/views/errors.py#L25-L29

Reply:
Sorry, you're right. I hadn't imported NoTokenException when I initially tested. I've updated so the exception is raised before form validation.